glfs.c

```diff
@@ -563,20 +563,24 @@ static int tcmu_glfs_open(struct tcmu_device *dev, bool reopen)
 		if (round_down(st.st_size, block_size) == dev_size)
 			goto out;
 
-		if (!reopen) {
-			ret = -EINVAL;
-			goto close;
-		}
+		if (!reopen)
+			tcmu_dev_warn(dev, "device size mismatch during initialization.\n");
+
+		tcmu_dev_warn(dev,
+			      "device size (%lld) and backing file size (%lld) not matching\n",
+			      (long long)dev_size, (long long) st.st_size);
 
 		/*
-		 * If we are here this should be in reopen path,
-		 * then we should also update the device size in
-		 * kernel.
+		 * Currently we cannot update the size to kernel during
+		 * initialization, because it will be overrided by kernel
+		 * after it get the genl reply. Here will just allow it
+		 * to load the device and we will run gluster-block tool
+		 * to update the size.
```

Comment:
I'm not sure this is correct.

In targetcli/rtslib we do:

1. make the device in configfs (do the mkdirs in /sys/kernel/config/core to setup user_N and whatever the device name is under that dir, etc).
2. Set dev_config, dev_size via the configfs control/info file.
3. enable the device (write 1 to the enable file in the device dir). This calls into the tcmu_configure_device callout for tcmu and this sends the TCMU_CMD_ADDED_DEVICE event.
4. runner handles that event by creating the device in the daemon and calling the handler->open callout.

Here if you set the tcmur_dev_update_size it will call into the kernel via the configfs control/info file and set dev_size.

5. runner returns NL response to tcmu.
6. tcmu returns response to lio core which returns to rtslib.
7. targetcli/rtslib then loops over the device attrs and writes to them assuming they are not the same as setup in the info/control file, which triggers the reconfig callout below for dev_size. In that callout you cannot update the size because you would hit the bug where it gets reset like you describe.

For the reconfig you would want to ignore the size if its not correct like you did below. But the dev_size should be the correct/updated value from step 4.

Reply:
> For the reconfig you would want to ignore the size if its not correct like you did below. But the dev_size should be the correct/updated value from step 4.

I mean the dev_size in tcmu-runner and the kernel will have been updated from step 4. The value being passed into the reconfig callout will be the old value from the /etc/target/saveconfig.json attributes array or wherever you are getting the storied attributes from.

Reply:
I try to catch some logs when restarting the tcmu-runner and gluster-blockd daemons at the same time:

```
tcmu-runner [INFO] handle_netlink:229: TCMU_CMD_REMOVED_DEVICE
tcmu-runner [INFO] handle_netlink:222: TCMU_CMD_ADDED_DEVICE, cfgstring = [tcm-user/0/block0/glfs/repvol@localhost/block-store/0f9483be-31a1-4ecb-8e3a-6137bedc6217]
tcmu-runner [INFO] tcmu_glfs_open:559 glfs/block0: device size (4294967296) and backing file size (5368709120)
tcmu-runner [INFO] handle_netlink:235: TCMU_CMD_RECONFIG_DEVICE
tcmu-runner [INFO] reconfig_device:163: TCMU_ATTR_DEV_CFG
tcmu-runner [INFO] handle_netlink:235: TCMU_CMD_RECONFIG_DEVICE
tcmu-runner [INFO] reconfig_device:172: TCMULIB_CFG_WRITE_CACHE
tcmu-runner [INFO] handle_netlink:235: TCMU_CMD_RECONFIG_DEVICE
tcmu-runner [INFO] reconfig_device:168: TCMULIB_CFG_DEV_SIZE, size = 4294967296

```
The dev_size could be set after the enable happens, see the last logs. And in the last step(my logs) above when setting the dev_size, it will call:

```
git diff drivers/target/target_core_user.c
diff --git a/drivers/target/target_core_user.c b/drivers/target/target_core_user.c
index 0b9dfa6..a53b011 100644
--- a/drivers/target/target_core_user.c
+++ b/drivers/target/target_core_user.c
@@ -2417,22 +2417,25 @@ static ssize_t tcmu_dev_size_store(struct config_item *item, const char *page,
        struct se_dev_attrib *da = container_of(to_config_group(item),
                                                struct se_dev_attrib, da_group);
        struct tcmu_dev *udev = TCMU_DEV(da->da_dev);
-       u64 val;
+       u64 val, old;
        int ret;
 
        ret = kstrtou64(page, 0, &val);
        if (ret < 0)
                return ret;
 
+       old = udev->dev_size;
+       udev->dev_size = val;
+
        /* Check if device has been configured before */
        if (target_dev_configured(&udev->se_dev)) {
                ret = tcmu_send_dev_size_event(udev, val);
                if (ret) {
                        pr_err("Unable to reconfigure device\n");
+                       udev->dev_size = old;
                        return ret;
                }
        }
-       udev->dev_size = val;
        return count;
 }
 CONFIGFS_ATTR(tcmu_, dev_size);
```
 As the code, the kernel/tcmu will send the dev_size reconfig NL request ---> tcmu-runner get the request and if we call `tcmur_dev_update_size(5G),` then the kernel/tcmu will update the udev->dev_size ---> tcmu-runner reply the NL request --> kernel/tcmu continue and will update` udev->dev_size = val;`, which the` val` here is the old size, 4G.

Reply:
@lxbsz 

There are 3 cases we are discussing right?

1. reconfig.

I agree it can set/reset the dev size after addition. As I said in the last comment that happens in step #7. We get the resize events because targetcli/rtslib loops over the attributes after the enablement.

That part of the patch is fine.

2. initial open (reopen=false):

I am saying that when the device is first added and runner calls handler->open() you can set the dev_size to the correct value and it will not get reset like in the reconfig case. 

After the initial open then we can hit #1, like I described in my last comment, but if you ignore like you are doing in your patch then that is fine.

3. reopen=true

We have a race where tcmu_glfs_reconfig could call tcmu_notify_conn_lost and in another thread glfs->open is called with reopen=true. If that thread calls tcmur_dev_update_size before the reconfig thread has returned the NL response then we hit issue 1.

So you could do the opposite in the open() call and we will use the correct size. See below for an example of what I mean.

You would need to still run targetlci or whatever tool to update the stored attribtutes in /etc/target/saveconfig.json.
```
@@ -558,20 +558,20 @@ static int tcmu_glfs_open(struct tcmu_device *dev, bool reopen)
                if (round_down(st.st_size, block_size) == dev_size)
                        goto out;

-               if (!reopen) {
-                       ret = -EINVAL;
-                       goto close;
-               }
+               tcmu_dev_warn(dev,
+                             "device size (%lld) and backing file size (%lld) not matching\n",
+                             (long long)dev_size, (long long) st.st_size);

                /*
-                * If we are here this should be in reopen path,
-                * then we should also update the device size in
-                * kernel.
+                * If reopen is processed while a reconfig we can race
+                * and the reconfig NL response may reset the size.
                 */
-               tcmu_dev_info(dev,
-                             "device size and backing size disagree:device %lld backing %lld\n",
-                             dev_size, (long long) st.st_size);
+               if (reopen)
+                       goto out;

+               tcmu_dev_warn(dev, "device size mismatch during initialization. Updating the kernel to  %lld. Please run \n",
+                            (long long) st.st_size);
+               /* Update the device size in kernel. */
                ret = tcmur_dev_update_size(dev, st.st_size);
                if (ret)
                        goto close;
@@ -762,10 +762,16 @@ static int tcmu_glfs_reconfig(struct tcmu_device *dev,
                        /* Let the targetcli command return success */
                        ret = 0;
                } else if (st.st_size != cfg->data.dev_size) {
-                       tcmu_dev_err(dev,
-                                    "device size and backing size disagree: device %"PRId64" backing %lld\n",
-                                    cfg->data.dev_size, (long long) st.st_size);
-                       ret = -EINVAL;
+                       /*
+                        * Currently we cannot update the size to kernel during
+                        * initialization, because it will be overrided by kernel
+                        * after it get the genl reply. Here will just allow it
+                        * to load the device and we will run gluster-block tool
+                        * to update the size.
+                        */
+                       tcmu_dev_warn(dev,
+                                     "device size (%lld) and backing file size (%lld) does not match\n",
+                                     (long long)cfg->data.dev_size, (long long) st.st_size);
                }
                return ret;
        case TCMULIB_CFG_DEV_CFGSTR:  
```

Reply:
>                 } else if (st.st_size != cfg->data.dev_size) {


One correction here. You need to add a

`                           ret = -EINVAL;
`


> -                       tcmu_dev_err(dev,
> -                                    "device size and backing size disagree: device %"PRId64" backing %lld\n",
> -                                    cfg->data.dev_size, (long long) st.st_size);
> -                       ret = -EINVAL;
> +                       /*
> +                        * Currently we cannot update the size to kernel during
> +                        * initialization, because it will be overrided by kernel
> +                        * after it get the genl reply. Here will just allow it
> +                        * to load the device and we will run gluster-block tool
> +                        * to update the size.
> +                        */
> +                       tcmu_dev_warn(dev,
> +                                     "device size (%lld) and backing file size (%lld) does not match\n",
> +                                     (long long)cfg->data.dev_size, (long long) st.st_size);
>                 }
>                 return ret;
>         case TCMULIB_CFG_DEV_CFGSTR:  
> ```

Reply:
> @lxbsz
> 
> There are 3 cases we are discussing right?
> 

Right.

> 1. reconfig.
> 
> I agree it can set/reset the dev size after addition. As I said in the last comment that happens in step # 7. We get the resize events because targetcli/rtslib loops over the attributes after the enablement.
> 
> That part of the patch is fine.
> 
> 1. initial open (reopen=false):
> 
> I am saying that when the device is first added and runner calls handler->open() you can set the dev_size to the correct value and it will not get reset like in the reconfig case.
> 
> After the initial open then we can hit # 1, like I described in my last comment, but if you ignore the bad size in the reconfig like you are doing in your patch then that is fine.
> 
> 1. reopen=true
> 
> We have a race where tcmu_glfs_reconfig could call tcmu_notify_conn_lost and in another thread glfs->open is called with reopen=true. If that thread calls tcmur_dev_update_size before the reconfig thread has returned the NL response then we hit issue 1.
> 
> So you could do the opposite in the open() call and we will use the correct size. See below for an example of what I mean.
> 
> You would need to still run targetlci or whatever tool to update the stored attribtutes in /etc/target/saveconfig.json.

Yeah, agree.

Updated it now. Thanks Mike.